### PR TITLE
Docs(spec): Add instance-partitioned code spec

### PR DESCRIPTION
**Pull request by tykeal**
## Summary

Spec-stage artifact for issue #734: `static_random` door codes are not
unique across Rental Control instances that share a parent lock, which
broke parent-lock programming in production and lets code-based guest
authentication resolve to the wrong booking.

Adds `specs/022-instance-code-partition/spec.md` plus its requirements
checklist. No implementation, no `plan.md`, no `tasks.md` — later
pipeline stages own those.

## What the spec covers

- Scope limited to the `static_random` generator; `date_based` and
  `last_four` are explicitly out of scope.
- Per-instance (not per-slot) partitioning of the code space, derived
  from each instance's existing `start_slot` / `max_events` measured
  against a shared parent-lock capacity constant, with no live
  cross-instance coordination.
- Capacity default of 250 plus an advanced override for large buildings
  using name-based virtual slots; the default case requires no new
  partition configuration when existing slot ranges are disjoint and in
  range.
- Deterministic intra-instance probing ordered primarily by reservation
  UID, with deterministic tie-breakers, and documented degraded fallback
  paths that always issue a code and warn without logging PIN values.
- Opt-out configuration option; partitioning is on by default, and opt-out
  is documented as forfeiting the cross-instance uniqueness guarantee.
- Determinism across restarts for generated candidates, and for returned
  codes when observed retained-code state is unchanged.
- Migration: readable observed lock PINs are retained; unreadable locks,
  duplicate retained PINs, and out-of-block retained PINs are explicit
  no-coordination exceptions.
- Explicit limitation: the current live collision will not self-heal;
  that needs the force-re-issue work in #735.

Refs #734. Also references #735, #736, and #741 as out-of-scope
follow-ups.

## Notes

No ambiguities required clarification markers; the design decisions were
settled by the maintainer and are recorded as requirements and
assumptions.

### CODE REVIEW

File: specs/022-instance-code-partition/spec.md [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+**Acceptance Scenarios**:
+
+1. **Given** a fresh installation or an upgrade with no new configuration,
+   **When** the `static_random` generator runs, **Then** partitioning is
+   active (opt-out, not opt-in).
+2. **Given** an operator who disables the option on an instance, **When**
+   codes are generated for that instance, **Then** the codes are exactly
+   those the previous whole-space generator produced for the same
+   bookings.
+3. **Given** the option is disabled on one instance only, **When** codes
```
[the remark this answers is not in the document]
Comment (tykeal) [reply specs/022-instance-code-partition/spec.md:220]: Fixed in 311ed01. The opt-out scenarios, edge cases, FR-021a, and FR-023b now require warning/confirmation and documentation that disabling partitioning restores whole-space generation and forfeits the cross-instance uniqueness guarantee for that instance.

File: specs/022-instance-code-partition/spec.md [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+  sensor-driven generation produce identical codes for the same booking.
+- **FR-003**: The system MUST NOT change the configured or default code
+  length as part of this feature.
+
+#### Block derivation
+
+- **FR-004**: The system MUST derive each instance's block of the code
+  space solely from that instance's own existing configuration — its
+  starting slot, its maximum event count, its code length — measured
+  against a parent-lock capacity value.
```
[the remark this answers is not in the document]
Comment (tykeal) [reply specs/022-instance-code-partition/spec.md:334]: Fixed in 311ed01. FR-004a through FR-004c now define the exact usable domain, half-open block formula, endpoint examples, and invalid out-of-range handling before disjointness is treated as testable.

File: specs/022-instance-code-partition/spec.md [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+#### Within-instance uniqueness
+
+- **FR-012**: The system MUST ensure that two bookings within the same
+  instance's plan never receive the same code, by deterministically
+  probing for the next free value within the instance's own block when a
+  candidate value is already taken.
+- **FR-013**: The probe MUST visit candidate values in a fixed, defined
+  order, and bookings MUST be resolved in a fixed order determined by
+  reservation UID, so that the outcome does not depend on calendar order,
+  fetch order, or processing order.
```
[the remark this answers is not in the document]
Comment (tykeal) [reply specs/022-instance-code-partition/spec.md]: Fixed in 311ed01. FR-013 now specifies the exact probe traversal, and FR-013a makes the booking resolution order total and stable by using UID as the primary key with deterministic tie-breakers for missing or repeated UIDs.

File: specs/022-instance-code-partition/spec.md [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+  captive-portal service resolves a code across integrations.
+
+## Success Criteria *(mandatory)*
+
+### Measurable Outcomes
+
+- **SC-001**: Across a simulated property of 10 instances sharing one
+  parent lock with adjacent slot ranges, 100% of generated
+  `static_random` codes are unique across the whole property, for every
+  tested booking set.
```
[the remark this answers is not in the document]
Comment (tykeal) [reply specs/022-instance-code-partition/spec.md]: Fixed in 311ed01. SC-001 is now scoped to partition-enabled instances with uniform capacity, seeded bookings, and sufficient block capacity, matching the documented degraded and opt-out exceptions rather than claiming an unconditional guarantee.

File: specs/022-instance-code-partition/spec.md [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+- **SC-004**: With partitioning disabled, 100% of generated codes match
+  the values produced by the previous release for the same bookings.
+- **SC-005**: A deployment left entirely at defaults obtains cross-
+  instance uniqueness with zero new configuration entered by the operator.
+- **SC-006**: No booking is ever left without a code: every degraded
+  condition (block exhausted, capacity larger than the code space, slot
+  range beyond capacity) still yields a correctly formatted code and emits
+  exactly one clear, instance-identifying warning.
+- **SC-007**: Upgrading an existing installation rotates zero codes for
+  bookings already written to the lock.
```
[the remark this answers is not in the document]
Comment (tykeal) [reply specs/022-instance-code-partition/spec.md]: Fixed in 311ed01. The migration text and SC-007 now qualify retention to active bookings whose existing lock PIN is readable during reconciliation, while preserving persistence as out of scope under #736.

File: specs/022-instance-code-partition/spec.md [hunk 6]
```diff
@@ -0,0 +1,511 @@
...
+duplicate broke programming on the parent while the child locks, which
+sync only their own subset, never noticed. Downstream consumers treat the
+door code as a credential and resolve a booking by code across *all*
+integrations, so a duplicate can authorize the wrong guest's booking.
+
+This feature makes cross-instance collisions structurally impossible for
+the `static_random` generator by partitioning the code space into disjoint
+per-instance blocks derived from configuration each instance already has,
+with no live coordination between instances, and resolves the remaining
+within-instance collisions with a deterministic probe.
```
[the remark this answers is not in the document]
Comment (tykeal) [reply specs/022-instance-code-partition/spec.md]: Fixed in 311ed01. The overview, User Story 1 test setup, edge cases, and SC-001 now scope the structural impossibility guarantee to valid enabled partitioning and explicitly mark degraded, opt-out, unseeded, and misconfigured paths as exceptions.

File: specs/022-instance-code-partition/spec.md [hunk 7]
```diff
@@ -0,0 +1,591 @@
...
+  fallback) mapped into the instance's block, and MUST NOT derive it from
+  the slot number the booking currently occupies, so that a reconciliation
+  slot move never changes a guest's code.
+
+#### Within-instance uniqueness
+
+- **FR-012**: The system MUST ensure that two bookings within the same
+  instance's plan never receive the same code, by deterministically
+  probing for the next free value within the instance's own block when a
+  candidate value is already taken.
```
[the remark this answers is not in the document]
Comment (tykeal) [reply specs/022-instance-code-partition/spec.md]: Fixed in 9bd8094. FR-012 is now scoped to non-degraded plans with enough free block values, while FR-015/FR-016 and SC-006 cover exhausted or empty-block fallback cases that may duplicate.

File: specs/022-instance-code-partition/spec.md [hunk 8]
```diff
@@ -0,0 +1,591 @@
...
+- **Capacity override applied unevenly across instances**
+  (misconfiguration, and the most likely way this feature silently
+  fails): capacity is the divisor that defines block boundaries, so
+  instances that disagree about it partition the same code space into
+  different-sized blocks, and those blocks can overlap even though the
+  slot ranges remain disjoint. Worked example in a 4-digit space of 9998
+  usable values — instance A on slots 10–30 with capacity 250 derives
+  roughly (399, 1199), while instance B on slots 40–60 with capacity 500
+  derives roughly (799, 1199): about 400 code values are shared and
+  cross-instance collisions return. This is a realistic path, not a
```
[the remark this answers is not in the document]
Comment (tykeal) [reply specs/022-instance-code-partition/spec.md]: Fixed in 9bd8094. The mismatched-capacity example now follows the FR-004b formula exactly: slots 10-30 at capacity 250 produce offsets [359,1199), slots 40-60 at capacity 500 produce [779,1199), for 420 overlapping values.

File: specs/022-instance-code-partition/spec.md [hunk 9]
```diff
@@ -0,0 +1,591 @@
...
+  offset at a time within the block, wrapping to the block start after the
+  block end, until a free value is found or the whole block has been
+  visited.
+- **FR-013a**: Bookings MUST be resolved in a total, stable order whose
+  primary key is reservation UID. Missing UIDs sort after present UIDs,
+  equal UIDs are broken by the legacy seed text, booking start time,
+  booking end time, and a stable source-calendar identifier, and any
+  remaining indistinguishable duplicate events are ordered by a canonical
+  serialization of those fields. The order MUST NOT depend on calendar
+  fetch order, entity iteration order, or processing order.
```
[the remark this answers is not in the document]
Comment (tykeal) [reply specs/022-instance-code-partition/spec.md]: Fixed in 9bd8094. FR-013a now requires immutable occurrence attributes, excludes mutable description text from equal-UID tie-breaking, and requires coalescing or warning for indistinguishable duplicate source events instead of relying on fetch order.

File: specs/022-instance-code-partition/spec.md [hunk 10]
```diff
@@ -0,0 +1,623 @@
...
+- **A booking is cancelled**: probe order may shift, so a *later* booking
+  that has not yet been written to the lock may receive a different code
+  than it would have. This is accepted; see Assumptions.
+- **Generators other than `static_random`**: `date_based` and `last_four`
+  are untouched and keep their present behaviour and their present
+  uniqueness weaknesses.
+- **No usable seed** (neither reservation UID nor description): the
+  existing fallback to the date-based generator applies unchanged, and the
+  resulting code is not partitioned. Cross-instance uniqueness is not
+  guaranteed for that booking.
```
[the remark this answers is not in the document]
Comment (tykeal) [reply specs/022-instance-code-partition/spec.md]: Fixed in 8710bb5. The no-UID/no-description path now warns per booking without logging the code, is listed as a degraded non-partitioned credential, and is included in FR-024/SC-006 coverage.

File: specs/022-instance-code-partition/spec.md [hunk 11]
```diff
@@ -0,0 +1,623 @@
...
+  that the instance will use whole-space generation and can duplicate
+  codes produced by sibling instances, forfeiting the cross-instance
+  uniqueness guarantee for that instance.
+- **FR-022**: The system MUST preserve existing code retention behaviour,
+  so that a code already observed on the lock for an active booking is
+  retained rather than rotated to the newly derived value.
+- **FR-022a**: Retained observed codes participate in the same plan-level
+  uniqueness calculation as newly generated codes. Retention wins for the
+  booking that already owns the observed PIN, and other bookings probe or
+  fall back rather than being assigned that same observed code.
```
[the remark this answers is not in the document]
Comment (tykeal) [reply specs/022-instance-code-partition/spec.md]: Fixed in 8710bb5. Out-of-block retained legacy/manual PINs are now an explicit retention/no-coordination exception to the cross-instance guarantee and SC-001; local plans still avoid observed retained PINs when assigning in-block generated codes.

File: specs/022-instance-code-partition/spec.md [hunk 12]
```diff
@@ -0,0 +1,663 @@
...
+
+**Independent Test**: Generate codes for a fixed set of bookings, discard
+all in-memory state, regenerate from the same inputs, and verify the
+mapping from booking to code is byte-identical.
+
+**Acceptance Scenarios**:
+
+1. **Given** an unchanged set of bookings and unchanged instance
+   configuration, **When** codes are generated repeatedly across restarts,
+   **Then** each booking receives the identical code every time.
```
[the remark this answers is not in the document]
Comment (tykeal) [reply specs/022-instance-code-partition/spec.md]: Fixed in 8f60464. User Story 2 and SC-003 now include identical observed retained-code state for returned-code determinism, while generated candidates remain deterministic from bookings and configuration alone.

File: specs/022-instance-code-partition/spec.md [hunk 13]
```diff
@@ -0,0 +1,663 @@
...
+
+## Success Criteria *(mandatory)*
+
+### Measurable Outcomes
+
+- **SC-001**: Across a simulated property of 10 partition-enabled
+  instances sharing one parent lock with adjacent slot ranges, a uniform
+  capacity, no unseeded bookings, no out-of-block retained PINs, and
+  sufficient block capacity, 100% of generated `static_random` codes are
+  unique across the whole property, for every tested booking set. This
```
[the remark this answers is not in the document]
Comment (tykeal) [reply specs/022-instance-code-partition/spec.md]: Fixed in 8f60464. SC-001 now excludes duplicate retained PINs and scopes the metric to newly generated static_random codes, matching the explicit retention exception in FR-022a.

File: specs/022-instance-code-partition/spec.md [hunk 14]
```diff
@@ -0,0 +1,677 @@
...
+  two source events remain indistinguishable after those keys, the system
+  MUST coalesce them as one booking for allocation or warn that their
+  relative order cannot be made deterministic; it MUST NOT silently rely
+  on fetch order, entity iteration order, or processing order.
+- **FR-014**: The probe MUST consider only codes within the instance's own
+  plan; it MUST NOT read codes belonging to other instances.
+- **FR-015**: When every value in the instance's block is taken, the
+  system MUST still return a code for the booking by falling back to the
+  unpartitioned whole-space value, and MUST record a warning identifying
+  the instance and the exhaustion condition.
```
[the remark this answers is not in the document]
Comment (tykeal) [reply specs/022-instance-code-partition/spec.md]: Fixed in fd365bc. FR-015a now applies the no-raw-PIN redaction requirement consistently to all partition-degradation warnings, including block exhaustion and retained-code exceptions; SC-006 also states warnings redact PIN values.

File: specs/022-instance-code-partition/spec.md [hunk 15]
```diff
@@ -0,0 +1,677 @@
...
+  a degraded, non-partitioned credential outside the cross-instance
+  uniqueness guarantee.
+
+#### Within-instance uniqueness
+
+- **FR-012**: For a non-degraded instance plan with enough free block
+  values, the system MUST ensure that two bookings within the same plan
+  never receive the same code, by deterministically probing for the next
+  free value within the instance's own block when a candidate value is
+  already taken.
```
[the remark this answers is not in the document]
Comment (tykeal) [reply specs/022-instance-code-partition/spec.md]: Fixed in fd365bc. FR-012 now excludes pre-existing duplicate retained PINs and scopes the distinct-code guarantee to newly generated and non-duplicate retained bookings, matching FR-022a and SC-002.

File: specs/022-instance-code-partition/spec.md [hunk 16]
```diff
@@ -0,0 +1,697 @@
...
+  a degraded, non-partitioned credential outside the cross-instance
+  uniqueness guarantee.
+
+#### Within-instance uniqueness
+
+- **FR-012**: For a non-degraded instance plan with enough free block
+  values and no pre-existing duplicate retained PINs, the system MUST
+  ensure that newly generated bookings and non-duplicate retained bookings
+  within the same plan never receive the same code, by deterministically
+  probing for the next free value within the instance's own block when a
```
[the remark this answers is not in the document]
Comment (tykeal) [reply specs/022-instance-code-partition/spec.md]: Fixed in 3d1f516. FR-012c now requires UID- or occurrence-aware allocation identity so different UIDs remain separate allocation participants even when legacy fingerprints match, and it defines how older retained PINs occupy the plan until rematched or expired.

File: specs/022-instance-code-partition/spec.md [hunk 17]
```diff
@@ -0,0 +1,712 @@
...
+
+#### Scope
+
+- **FR-001**: The system MUST apply code-space partitioning only to the
+  `static_random` generator. The `date_based` and `last_four` generators
+  MUST be left byte-for-byte unchanged.
+- **FR-002**: The system MUST expose one behaviourally identical
+  partitioned allocation contract to every generation path, so that
+  coordinator-driven and sensor-driven generation produce identical codes
+  for the same booking set and observed retention state.
```
[the remark this answers is not in the document]
Comment (tykeal) [reply specs/022-instance-code-partition/spec.md]: Fixed in 6fedddc. FR-002 is now scoped to every static_random generation path, preserving FR-001 exclusions for date_based and last_four. FR-006 and FR-011a were also tightened to the same in-block static_random contract.

File: specs/022-instance-code-partition/spec.md [hunk 18]
```diff
@@ -0,0 +1,721 @@
...
+  no free value remains in the block, the generator falls back to the
+  unpartitioned whole-space value for that booking and records a clear
+  warning naming the instance and advising a longer code length or a lower
+  capacity only when the lower value is the real shared parent-lock
+  capacity and is applied uniformly to every sibling instance. The warning
+  is recorded once for each affected booking, because each fallback
+  credential can collide independently. Issuing a
+  possibly-duplicate code is preferred over issuing none, because a
+  missing code locks a guest out immediately whereas a duplicate is rare
+  and detectable.
```
[the remark this answers is not in the document]
Comment (tykeal) [reply specs/022-instance-code-partition/spec.md]: Fixed in 6307786. The exhaustion text no longer promises duplicate detectability; it now states the residual collision and ambiguous-authentication risk. FR-015 also pins fallback to the legacy static_random whole-space value.

File: specs/022-instance-code-partition/spec.md [hunk 19]
```diff
@@ -0,0 +1,721 @@
...
+  bookings for the instance after the integration's normal calendar
+  parsing, checkout, cancellation, and maximum-event filtering, plus every
+  slot-eligible retained booking whose managed slot still has an observed
+  PIN, including future retained bookings and retained ghost placeholders
+  that temporarily represent missing feed events with an observed PIN.
+  Overflow bookings that will not be assigned a slot in the current plan,
+  cancelled bookings, checked-out bookings, ghost placeholders without an
+  observed PIN, and bookings outside the instance's managed slot range do
+  not participate. When no current plan exists, such as the first refresh,
+  both paths MUST build allocation from that same population before
```
[the remark this answers is not in the document]
Comment (tykeal) [reply specs/022-instance-code-partition/spec.md]: Fixed in 6307786. FR-012a now keeps locally observed stale PINs for cancelled, checked-out, retained overflow, and ghost cases in the occupied-code set until the clear is confirmed.

File: specs/022-instance-code-partition/spec.md [hunk 20]
```diff
@@ -0,0 +1,721 @@
...
+  booking that already owns the observed PIN, and other in-block bookings
+  probe rather than being assigned that same observed code. Pre-existing
+  duplicate observed PINs are retained for their bookings as an explicit
+  exception; new in-block allocations avoid the locally observed PIN, but
+  block-exhaustion fallback can still duplicate and must warn.
+- **FR-022b**: During upgrade and later reconciliation, any readable
+  observed PIN in this instance's managed slots for a slot-eligible
+  booking MUST be retained for that booking before comparing against newly
+  generated partitioned or legacy values. This includes readable
+  pre-feature PINs that happen to equal the old whole-space generator
```
[the remark this answers is not in the document]
Comment (tykeal) [reply specs/022-instance-code-partition/spec.md]: Fixed in 6307786. FR-022b is now limited to the first migration reconciliation after enabling partitioned generation, while FR-022c leaves normal explicit reissue/update semantics in place after that window.

File: specs/022-instance-code-partition/spec.md [hunk 21]
```diff
@@ -0,0 +1,737 @@
...
+  partitioned generation, any readable observed PIN in this instance's
+  managed slots for a slot-eligible booking MUST be retained for that
+  booking before comparing against newly generated partitioned or legacy
+  values. This includes readable pre-feature PINs that happen to equal the
+  old whole-space generator output.
+- **FR-022c**: On the migration from pre-partitioned generation to
+  partitioned generation, FR-022b overrides any "update generated code"
+  setting for readable observed PINs so existing guest credentials do not
+  rotate solely because the generator changed. Outside that migration
+  window, existing explicit reissue/update semantics continue to control
```
[the remark this answers is not in the document]
Comment (tykeal) [reply specs/022-instance-code-partition/spec.md]: Fixed in fab97b1. FR-022c now defines the migration window using a stored partitioning schema version on the config entry: active while missing/older, cleared only after one successful reconciliation records the current version.

**Comment by tykeal** [comment]
Closing unmerged. This spec describes per-instance code-space partitioning, which was a workaround for the absence of shared state between Rental Control config entries. The maintainer has decided to pursue a shared allocator with a persisted registry instead (see the new architecture issue), which supersedes this approach. Superseded, not rejected: the analysis here — particularly the requirement for uniform parent-lock capacity and the sensor/coordinator parity problem — informed the new design.
